product_supplier_code_purchase/models/purchase_order.py

```diff
@@ -1,27 +1,64 @@
-# -*- coding: utf-8 -*-
-# Copyright 2015-17 Eficent Business and IT Consulting Services, S.L.
-#           (http://www.eficent.com)
+# Copyright 2015-17 ForgeFlow, S.L.
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
-
 from odoo import api, fields, models
+from odoo.exceptions import AccessError
 
 
 class PurchaseOrderLine(models.Model):
     _inherit = 'purchase.order.line'
 
-    product_supplier_code = fields.Char(
-        string='Product Supplier Code',
-        compute='_compute_product_supplier_code'
-    )
+    product_supplier_code = fields.Char(string='Product Supplier Code')
 
     @api.multi
-    @api.depends('product_id.seller_ids.name', 'partner_id',
-                 'product_id.seller_ids.product_code')
-    def _compute_product_supplier_code(self):
+    @api.onchange(
+        'partner_id',
+        'product_id',
+    )
+    def _onchange_product_code(self):
         for line in self:
             supplier_info = line.product_id.seller_ids.filtered(
-                lambda s: (s.product_id == line.product_id and
-                           s.name == line.partner_id))
+                lambda s: (
+                    s.product_id == line.product_id
+                    and s.name == line.partner_id
+                )
+            )
             if supplier_info:
                 code = supplier_info[0].product_code or ''
                 line.product_supplier_code = code
+            else:
+                supplier_info = line.product_id.seller_ids.filtered(
+                    lambda s: (
+                        s.product_tmpl_id == line.product_id.product_tmpl_id
+                        and s.name == line.partner_id
+                    )
+                )
+                if supplier_info:
+                    code = supplier_info[0].product_code or ''
+                    line.product_supplier_code = code
+
+
+class PurchaseOrder(models.Model):
+    _inherit = 'purchase.order'
+
+    @api.multi
+    def _add_supplier_to_product(self):
+        super()._add_supplier_to_product()
+        for line in self.order_line:
+            partner = (
+                self.partner_id
+                if not self.partner_id.parent_id
+                else self.partner_id.parent_id
+            )
+            if partner in line.product_id.seller_ids.mapped('name'):
+                seller = line.product_id._select_seller(
+                    partner_id=line.partner_id,
+                    quantity=line.product_qty,
+                    date=line.order_id.date_order
+                    and line.order_id.date_order.date(),
+                    uom_id=line.product_uom,
+                )
+                if seller:
+                    try:
+                        seller['product_code'] = line.product_supplier_code
+                    except AccessError:
```

Comment:
I don't understand such try / catch AccessError. Could you explain this point ?

Reply:
Odoo does the same thing when updating supplierinfo: https://github.com/odoo/odoo/blob/12.0/addons/purchase/models/purchase.py#L396

I guess this is when you have permissions in PO lines but not in supplierinfo. I did not get any error there, but just in case, I prefer to follow what Odoo does.